added keybindings for syntax test development

keith-hall opened the pull request:
added keybindings and a visual helper for syntax test development as referenced to from https://github.com/SublimeText/PackageDev/issues/71#issuecomment-267342418

jcberquist commented:
@keith-hall: This is minor, but would you consider adjusting the highlight listener to not erase the `current_syntax_test` region if it is unchanged? Currently, as the cursor moves on the test line, the highlight region flickers as it is erased and redrawn.

jcberquist commented:
@keith-hall: Thanks! Am I right that you don't actually need to erase the region explicitly if you call `view.add_regions()` again with the same key? I think it overwrites what is already there?

keith-hall commented:
great idea @jcberquist - and thanks for the tip! it works perfectly and saved me lots of time ;)

keith-hall commented:
out of interest, does anyone ever really work with multiple cursors while writing syntax tests? I don't, and so I've only added support for the 1st cursor in this PR so far.

FichteFoll commented on syntax-test_dev.py:
```diff
@@ -0,0 +1,271 @@
+            col_end += 1
+
+        region = sublime.Region(line.begin() + col_start, line.begin() + col_end)
+        self.view.add_regions('current_syntax_test', [region], 'text', '', sublime.DRAW_NO_FILL)
```
Imo this should be configurable. Both the style and the scope name.
I already have highlighting from https://packagecontrol.io/packages/SublimeLinter-contrib-sublime-syntax on test lines, so they *might* conflict (although SL's coloring is almost always red).

FichteFoll commented on syntax-test_dev.py:
```diff
@@ -0,0 +1,271 @@
+class SyntaxTestEventListener(sublime_plugin.EventListener):
+    """Event Listener for responding to keybinding contexts"""
+
+    def on_query_context(self, view, key, operator, operand, match_all):
```
Is `on_query_context` not active in ViewEventListeners?

keith-hall replied:
I'm sure when I first tried it, it wasn't... but seems to work now...

FichteFoll commented on syntax-test_dev.py:
```diff
@@ -0,0 +1,271 @@
+import re
+
+
+def get_syntax_test_tokens(view):
```
Since all "helper" functions revolve around a given view,
it could make sense architecture-wise to group the functions within a class and store instances of these in a way that the plugin interface points (commands and event listeners) can access it.

Just a thought. I don't know if this would end up being better.

FichteFoll commented:
>out of interest, does anyone ever really work with multiple cursors while writing syntax tests?

I do occasionally, but mostly for aligning which doesn't conflict with this feature.

keith-hall commented on syntax-test_dev.py:
```diff
@@ -0,0 +1,271 @@
+        """
+
+        view = self.view
+        char = args.get('character', '^')
```
I did this for consistency when typing the `-` after `<`, i.e. to test the scope at the comment marker

keith-hall commented on syntax-test_dev.py:
```diff
@@ -0,0 +1,271 @@
+        def file_contains_syntax_tests():
+            return is_syntax_test_file(view)
+
+        if key not in locals():
```
thanks for the info - I don't seem to get notifications about comments on my gists.

FichteFoll replied:
Yeah, unfortunately you don't. And you can't even mention people iirc.

keith-hall commented:
I've made some improvements as suggested, thanks :) I will probably only have time to make more after the weekend, but if I can tomorrow, I will try :)

FichteFoll commented:
Just a note that I'll most likely be focussing my time on writing the Package Control channel review tool before working on PackageDev.

FichteFoll commented on syntax-test_dev.py:
```diff
@@ -44,6 +41,10 @@ def get_details_of_test_assertion_line(view, pos):
     - the assertion characters (2nd item in tuple aka `assertion_colrange`)
     """
 
+    AssertionLineDetails = namedtuple(
```
I preferred this in the module root (since its instances are being used outside of the function).

FichteFoll commented on PackageDev.sublime-settings:
```diff
@@ -0,0 +1,12 @@
+    // "DRAW_EMPTY", "HIDE_ON_MINIMAP", "DRAW_EMPTY_AS_OVERWRITE", "DRAW_NO_FILL",
+    // "DRAW_NO_OUTLINE", "DRAW_SOLID_UNDERLINE", "DRAW_STIPPLED_UNDERLINE",
+    // "DRAW_SQUIGGLY_UNDERLINE", "HIDDEN", "PERSISTENT"
+    "styles_for_syntax_test_highlight": ["DRAW_NO_FILL"],
```
I prefer `syntax_test_highlight_styles`. Same for `scope`.

FichteFoll commented on syntax-test_dev.py:
```diff
@@ -236,11 +237,30 @@ def on_selection_modified_async(self):
+        #     ]:
+        #         style_flags |= getattr(sublime, style)
+        for style in styles:
+            # check the style matches the casing and naming used for constants in `sublime.py`
```
I prefer the explicit check here since the module contains other constants that would not make sense in this context. I know that you hardly ever encounter a situation where a user inserts this in their settings, but I like to be on the safe side when evaluating user input, always.

keith-hall commented:
I believe I have addressed all the points you raised in your review, @FichteFoll.

I've also got it to suggest partial scopes:
![ezgif com-8d856ff244](https://cloud.githubusercontent.com/assets/11882719/21346302/87eb1954-c6ac-11e6-9c2c-db19ba842d51.gif)

and cope with comment end markers:
![ezgif com-122ca4c117](https://cloud.githubusercontent.com/assets/11882719/21346713/3f48ed46-c6ae-11e6-80c7-a3c80ade229f.gif)

And proved it works nicely with line comments that also scope the trailing newline:
![ezgif com-710925ebe9](https://cloud.githubusercontent.com/assets/11882719/21346774/82b68e44-c6ae-11e6-96f0-ef29d40a1207.gif)

FichteFoll commented:
Looks awesome. Thanks for addressing my feedback so quickly. I intend to do more work with this package before the end of the year (holidays, yay), and will hopefully merge this and work on other parts that have been bothering me for a long while.
